experimental/uefi/channel/src/lib.rs

```diff
@@ -66,16 +118,205 @@ impl Frame {
+            0 => vec![vec![]],
+            _ => message
+                .body
+                .chunks(MAX_FRAME_BODY_SIZE)
+                // TODO(#2848): It'd be nice if we didn't have to reallocate here.
+                // We may be able to avoid doing so by making [`Frame`] use
+                // slices and lifetimes. Or alternatively use the Bytes crate for
+                // reference counting.
+                .map(|frame_body| frame_body.to_vec())
+                .collect(),
+        };
+        let length = frame_bodies.len();
+        frame_bodies
+            .into_iter()
+            .enumerate()
+            .map(|(index, body)| {
+                let flag = {
+                    if index == 0 {
+                        match length {
+                            1 => Flag::AtomicFrame,
+                            _ => Flag::StreamStart,
+                        }
+                    } else if index == length - 1 {
+                        Flag::StreamEnd
+                    } else {
+                        Flag::StreamContinuance
+                    }
+                };
+                Frame {
+                    method_or_status: message.method_or_status,
+                    flag,
+                    body,
+                }
+            })
+            .collect()
+    }
+}
+
+/// Struct to incrementally build a [`Message`] from a set of [`Frame`]s.
+#[derive(Default, Debug, PartialEq, Eq)]
+struct PartialMessage {
+    inner: Option<Message>,
+}
+#[derive(Debug, PartialEq, Eq)]
+enum MessageReconstructionErrors {
+    ExpectedStartFrame,
+    DoubleStartFrame,
+    MismatchingFrameHeader,
+}
+impl MessageReconstructionErrors {
+    fn as_str(&self) -> &'static str {
+        match self {
+            MessageReconstructionErrors::ExpectedStartFrame => "Invalid frame order. The first frame must be a start frame.",
+            MessageReconstructionErrors::DoubleStartFrame => "Invalid frame order. The frame method/status header does must be the same for all frames belonging to the same message.",
+            MessageReconstructionErrors::MismatchingFrameHeader => "Invalid frame order. Already received a start frame for this message."
+        }
+    }
+}
+impl From<MessageReconstructionErrors> for anyhow::Error {
+    fn from(error: MessageReconstructionErrors) -> Self {
+        anyhow::Error::msg(error.as_str())
+    }
+}
+impl PartialMessage {
+    /// Attempt to complete the [`PartialMessage`] with a frame. Returns
+    /// an outer [`Result`] to handle errors resulting from invalid arguments.
+    /// Returns an inner [`Result`] that contains either the succesfully
+    /// completed message, or if more frames are needed to complete the message,
+    /// a new [`PartialMessage`] that includes the previously passed frames.
+    pub fn try_complete(
+        self,
+        mut frame: Frame,
+    ) -> Result<Result<Message, Self>, MessageReconstructionErrors> {
```

Comment:
+1, I think the current return type is a bit unusual and not super ergonomic.

Reply:
RE the result return type: that's fair, I was going back and forth between whether to use an Enum (ala `Enum { Complete(Message), Incomplete(PartialMessage) }` or a result. :) 

RE recursion though, that actually comes from the fact that this approach is more functional, in so far as that it does not use mutation. 

Expressed as a trait, the current version looks something like:
```
trait FunctionalAssembler: Sized {
    fn try_complete(
        self,
        frame: Frame,
    ) -> Result<Result<Message, Self>, MessageReconstructionErrors>;
}
```

A mutable (and by extension iteration-friendly) struct would look something like this:

```
trait MessageBuilder {
    fn requires_more_frames(&self) -> bool;
    fn get_completed_message(self) -> Message;
    fn add_frame(&mut self, frame: Frame) -> Result<(), MessageReconstructionErrors>;
}
```

We could then iterate over it via:

```
while message_builder.requires_more_frames() {
    let frame = framed.read_frame()?;
    message_builder.add_frame(frame)?;
}
let reconstructed_message = message_builder.get_completed_message();
```

While friendly on the surface, I think it actually a bit of complexity, since it methods to align in behavior. (Eg if requires_more_frames returns true, add_frame should fail. If requires_more_frames is false, get_completed_message is still available but should throw). It feels more ergonomic, but in each state there's methods that while available will need to error if called.

We used the iterative approach before, and arguably it led to some unwelcome coupling: https://github.com/project-oak/oak/pull/2906/files/e613d85fbb2be9963a71caa075962e55609fd5e0#r884977491

Hence I'm personally partial to the function approach. But happy to go with iterative if people prefer. :)

Reply:
Definitely happy to change the inner Result to an Enum though either way :)

Reply:
Also should note that the need for recursion should go away once we implement message multiplexing :)

Reply:
I disagree that the recursive approach is stateless. It hides the state in the execution stack and does require mutation (`append_partial_message`).

I agree that the recursion will go away once interleaving of messages is implemented, which is part of the reason I preferred an approach that would be closer to the future state. Having said that, I don't have a strong objection to temporarily keeping this approach if it unblocks other work.

Reply:
Agreed that it's about hiding inner mutability & state from the outside, it doesn't go away entirely. Not sure if mutability necessarily makes the result closer to what we'll have with message interleaving, but ok to go with it.

Reply:
I assume interleaving will be implmented using some cache (e.g. LRU map). When a frame is received the message ID will be the key for finding the partial buffer in the cache. If the in-progress buffer is found (and the flags are OK0 it will be mutated by appending the bytes from the latest frame. This seems much closer to the iterative approach than the recursive approach.

Perhaps there is a different implementation for interleaving that will look closer to the recursive approach, but I could not think of an abvious candidate.

Reply:
In the recursive approach, instead of appending the partial buffer in the LRU cache in place, it would be replaced with an appended version of the same buffer.